File: MiniDOMOffice/PBX/RegistroChiamate/StoricoHandlers/StoricoHandlerCR.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Office;
using static minidom.CustomerCalls;

namespace minidom
{
    public partial class Office
    {

        /// <summary>
        /// Handler delle chiamate registrate
        /// </summary>
        /// <remarks></remarks>
        public class StoricoHandlerCR 
            : StoricoHandlerBase
        {

            /// <summary>
            /// Costruttore
            /// </summary>
            public StoricoHandlerCR()
            {
            }

            /// <summary>
            /// Aggiunge le chiamate registrate
            /// </summary>
            /// <param name="items"></param>
            /// <param name="filter"></param>
            protected override void AggiungiInternal(
                                    CCollection<StoricoAction> items, 
                                    CRMFindFilter filter
                                    )
            {
                int cnt = 0;

                using (var cursor = new ChiamataRegistrataCursor())
                {
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    if (filter.Dal.HasValue)
                    {
                        cursor.DataInizio.Value = filter.Dal.Value;
                        cursor.DataInizio.Operator = OP.OP_GT;
                        if (filter.Al.HasValue)
                        {
                            cursor.DataInizio.Value1 = filter.Al.Value;
                            cursor.DataInizio.Operator = OP.OP_BETWEEN;
                        }
                    }
                    else if (filter.Al.HasValue)
                    {
                        cursor.DataInizio.Value = filter.Al.Value;
                        cursor.DataInizio.Operator = OP.OP_LE;
                    }

                    if (!string.IsNullOrEmpty(filter.Contenuto))
                    {
                        // cursor..Value = filter.Contenuto & "%"
                        // cursor.Note.Operator = OP.OP_LIKE
                    }
                    // If filter.DettaglioEsito Then
                    // filter.etichetta
                    // If filter.IDOperatore <> 0 Then cursor.IDOperator.Value = filter.IDOperatore
                    // If filter.IDPuntoOperativo <> 0 Then cursor.IDPuntoOperativo.Value = filter.IDPuntoOperativo
                    if (filter.IDPersona != 0)
                    {
                        cursor.IDChiamato.Value = filter.IDPersona;
                    }
                    else if (!string.IsNullOrEmpty(filter.Nominativo))
                    {
                        cursor.NomeChiamato.Value = filter.Nominativo + "%";
                        cursor.NomeChiamato.Operator = OP.OP_LIKE;
                    }

                    if (!string.IsNullOrEmpty(filter.Numero))
                    {
                        cursor.ANumero.Value = filter.Numero + "%";
                        cursor.ANumero.Operator = OP.OP_LIKE;
                    }

                    if (filter.StatoConversazione.HasValue)
                    {
                        switch (filter.StatoConversazione)
                        {
                            case 0: // In Attesa
                                {
                                    cursor.StatoChiamata.ValueIn(new object[] { StatoChiamataRegistrata.Sconosciuto, StatoChiamataRegistrata.Composizione });
                                    break;
                                }

                            case 1: // in corso
                                {
                                    cursor.StatoChiamata.ValueIn(new object[] { StatoChiamataRegistrata.InCorso });
                                    break;
                                }

                            default:
                                {
                                    cursor.StatoChiamata.ValueIn(new object[] { StatoChiamataRegistrata.AgganciatoChiamante, StatoChiamataRegistrata.AgganciatoChiamato, StatoChiamataRegistrata.Errore, StatoChiamataRegistrata.NonRisposto, StatoChiamataRegistrata.Rifiutata });
                                    break;
                                }
                        }
                    }

                    if (filter.IDContesto.HasValue)
                    {
                        // cursor2.ContextType.Value = filter.TipoContesto
                        // cursor2.ContextID.Value = filter.IDContesto
                    }

                    if (!string.IsNullOrEmpty(filter.Scopo))
                    {
                    }

                    cursor.DataInizio.SortOrder = SortEnum.SORT_DESC;
                    cursor.IgnoreRights = filter.IgnoreRights;
                    // cursor2.StatoOperazione.Value = StatoRichiestaCERQ.RITIRATA
                    while (cursor.Read() && (!filter.nMax.HasValue || cnt < filter.nMax))
                    {
                        cnt += 1;
                        // items.Add(cursor2.Item)
                        AddActivities(items, cursor.Item);
                    }
                }
            }

            private void AddActivities(CCollection<CustomerCalls.StoricoAction> col, ChiamataRegistrata c)
            {
                var action = new CustomerCalls.StoricoAction();
                action.Data = c.DataInizio;
                action.IDOperatore = DBUtils.GetID(Sistema.Users.KnownUsers.SystemUser, 0);
                action.NomeOperatore = Sistema.Users.KnownUsers.SystemUser.Nominativo;
                action.IDCliente = c.IDChiamato;
                action.NomeCliente = c.NomeChiamato;
                action.Note = "Chiamata da: " + Sistema.Formats.FormatNumber(c.DaNumero) + " a " + Sistema.Formats.FormatNumber(c.ANumero) + DMD.Strings.vbCrLf + "Esito: " + Enum.GetName(typeof(EsitoChiamataRegistrata), c.EsitoChiamata) + "Inizio: " + Sistema.Formats.FormatUserDateTime(c.DataInizio);
                if (c.DataRisposta.HasValue)
                    action.Note += DMD.Strings.vbCrLf + "Risposta: " + Sistema.Formats.FormatUserDateTime(c.DataRisposta);
                if (c.DataRisposta.HasValue && c.DataInizio.HasValue)
                {
                    action.Note += " (Attesa: " + Sistema.Formats.FormatDurata((long?)(c.DataRisposta.Value - c.DataInizio.Value).TotalSeconds) + ")";
                    action.Attesa = 0d;
                }

                if (c.DataFine.HasValue)
                {
                    action.Note += DMD.Strings.vbCrLf + "Fine: " + Sistema.Formats.FormatUserDateTime(c.DataFine);
                    if (c.DataRisposta.HasValue)
                    {
                        action.Durata = (c.DataFine.Value - c.DataInizio.Value).TotalSeconds;
                        action.Attesa = (c.DataRisposta.Value - c.DataInizio.Value).TotalSeconds;
                        action.Note += " (Durata conversazione: " + Sistema.Formats.FormatDurata((long?)(c.DataFine.Value - c.DataRisposta.Value).TotalSeconds) + ")";
                    }
                    else if (c.DataInizio.HasValue)
                    {
                        action.Durata = (c.DataFine.Value - c.DataInizio.Value).TotalSeconds;
                        action.Attesa = action.Durata;
                        action.Note += " (Attesa: " + Sistema.Formats.FormatDurata((long?)action.Durata) + ")";
                    }
                }

                action.Scopo = "";
                action.NumeroOIndirizzo = c.ANumero;
                switch (c.EsitoChiamata)
                {
                    case EsitoChiamataRegistrata.Risposto:
                        {
                            action.Esito = CustomerCalls.EsitoChiamata.OK;
                            break;
                        }

                    case EsitoChiamataRegistrata.NonRisposto:
                        {
                            action.Esito = CustomerCalls.EsitoChiamata.NESSUNA_RISPOSTA;
                            break;
                        }
                }

                action.DettaglioEsito = c.EsitoChiamataEx;
                action.Tag = c;
                action.Ricevuta = false;
                switch (c.StatoChiamata)
                {
                    case StatoChiamataRegistrata.Sconosciuto:
                    case StatoChiamataRegistrata.InCorso:
                    case StatoChiamataRegistrata.Composizione:
                        {
                            action.StatoConversazione = CustomerCalls.StatoConversazione.INATTESA;
                            break;
                        }

                    default:
                        {
                            action.StatoConversazione = CustomerCalls.StatoConversazione.CONCLUSO;
                            break;
                        }
                }

                col.Add(action);
            }

            /// <summary>
            /// Tipi supportati
            /// </summary>
            /// <param name="items"></param>
            protected override void FillSupportedTypes(CKeyCollection<string> items)
            {
                items.Add("ChiamataRegistrata", "Chiamata Registrata");
            }
        }
    }
}

File: MiniDOMOffice/PBX/RegistroChiamate/StoricoHandlerCR.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases
Imports minidom.CustomerCalls
Imports minidom.Anagrafica



Partial Public Class Office

    ''' <summary>
    ''' Handler delle chiamate registrate
    ''' </summary>
    ''' <remarks></remarks>
    Public Class StoricoHandlerCR
        Inherits StoricoHandlerBase

        Public Sub New()

        End Sub

        Protected Overrides Sub AggiungiInternal(items As CCollection(Of StoricoAction), filter As CRMFindFilter)
            Dim cnt As Integer = 0
            Dim cursor As New ChiamataRegistrataCursor

#If Not DEBUG Then
            Try
#End If
            cursor.Stato.Value = ObjectStatus.OBJECT_VALID
            If filter.Dal.HasValue Then
                cursor.DataInizio.Value = filter.Dal.Value
                cursor.DataInizio.Operator = OP.OP_GT
                If filter.Al.HasValue Then
                    cursor.DataInizio.Value1 = filter.Al.Value
                    cursor.DataInizio.Operator = OP.OP_BETWEEN
                End If
            ElseIf filter.Al.HasValue Then
                cursor.DataInizio.Value = filter.Al.Value
                cursor.DataInizio.Operator = OP.OP_LE
            End If
            If filter.Contenuto <> "" Then
                'cursor..Value = filter.Contenuto & "%"
                'cursor.Note.Operator = OP.OP_LIKE
            End If
            'If filter.DettaglioEsito Then
            'filter.etichetta
            'If filter.IDOperatore <> 0 Then cursor.IDOperator.Value = filter.IDOperatore
            'If filter.IDPuntoOperativo <> 0 Then cursor.IDPuntoOperativo.Value = filter.IDPuntoOperativo
            If (filter.IDPersona <> 0) Then
                cursor.IDChiamato.Value = filter.IDPersona
            ElseIf filter.Nominativo <> "" Then
                cursor.NomeChiamato.Value = filter.Nominativo & "%"
                cursor.NomeChiamato.Operator = OP.OP_LIKE
            End If
            If filter.Numero <> "" Then
                cursor.ANumero.Value = filter.Numero & "%"
                cursor.ANumero.Operator = OP.OP_LIKE
            End If
            If filter.StatoConversazione.HasValue Then
                Select Case filter.StatoConversazione
                    Case 0 'In Attesa
                        cursor.StatoChiamata.ValueIn(New Object() {StatoChiamataRegistrata.Sconosciuto, StatoChiamataRegistrata.Composizione})
                    Case 1 'in corso
                        cursor.StatoChiamata.ValueIn(New Object() {StatoChiamataRegistrata.InCorso})
                    Case Else
                        cursor.StatoChiamata.ValueIn(New Object() {StatoChiamataRegistrata.AgganciatoChiamante, StatoChiamataRegistrata.AgganciatoChiamato, StatoChiamataRegistrata.Errore, StatoChiamataRegistrata.NonRisposto, StatoChiamataRegistrata.Rifiutata})
                End Select
            End If
            If (filter.IDContesto.HasValue) Then
                'cursor2.ContextType.Value = filter.TipoContesto
                'cursor2.ContextID.Value = filter.IDContesto
            End If
            If (filter.Scopo <> "") Then

            End If

            cursor.DataInizio.SortOrder = SortEnum.SORT_DESC
            cursor.IgnoreRights = filter.IgnoreRights
            'cursor2.StatoOperazione.Value = StatoRichiestaCERQ.RITIRATA
            While Not cursor.EOF AndAlso (Not filter.nMax.HasValue OrElse cnt < filter.nMax)
                cnt += 1
                'items.Add(cursor2.Item)
                Me.AddActivities(items, cursor.Item)
                cursor.MoveNext()
            End While
#If Not DEBUG Then
            Catch ex As Exception
                Throw
            Finally
#End If
            cursor.Dispose()
#If Not DEBUG Then
            End Try
#End If
        End Sub



        Private Sub AddActivities(ByVal col As CCollection(Of StoricoAction), ByVal c As ChiamataRegistrata)
            Dim action As New StoricoAction

            action.Data = c.DataInizio
            action.IDOperatore = GetID(Sistema.Users.KnownUsers.SystemUser)
            action.NomeOperatore = Sistema.Users.KnownUsers.SystemUser.Nominativo
            action.IDCliente = c.IDChiamato
            action.NomeCliente = c.NomeChiamato
            action.Note = "Chiamata da: " & Formats.FormatNumber(c.DaNumero) & " a " & Formats.FormatNumber(c.ANumero) & vbCrLf &
                          "Esito: " & [Enum].GetName(GetType(EsitoChiamataRegistrata), c.EsitoChiamata) &
                          "Inizio: " & Formats.FormatUserDateTime(c.DataInizio)
            If (c.DataRisposta.HasValue) Then action.Note &= vbCrLf & "Risposta: " & Formats.FormatUserDateTime(c.DataRisposta)
            If (c.DataRisposta.HasValue AndAlso c.DataInizio.HasValue) Then
                action.Note &= " (Attesa: " & Formats.FormatDurata((c.DataRisposta.Value - c.DataInizio.Value).TotalSeconds) & ")"
                action.Attesa = 0
            End If
            If (c.DataFine.HasValue) Then
                action.Note &= vbCrLf & "Fine: " & Formats.FormatUserDateTime(c.DataFine)
                If (c.DataRisposta.HasValue) Then
                    action.Durata = (c.DataFine.Value - c.DataInizio.Value).TotalSeconds
                    action.Attesa = (c.DataRisposta.Value - c.DataInizio.Value).TotalSeconds
                    action.Note &= " (Durata conversazione: " & Formats.FormatDurata((c.DataFine.Value - c.DataRisposta.Value).TotalSeconds) & ")"

                ElseIf (c.DataInizio.HasValue) Then
                    action.Durata = (c.DataFine.Value - c.DataInizio.Value).TotalSeconds
                    action.Attesa = action.Durata
                    action.Note &= " (Attesa: " & Formats.FormatDurata(action.Durata) & ")"
                End If

            End If
            action.Scopo = ""
            action.NumeroOIndirizzo = c.ANumero
            Select Case c.EsitoChiamata
                Case EsitoChiamataRegistrata.NonRisposto
                    action.Esito = EsitoChiamata.RIPOSTA
                Case EsitoChiamataRegistrata.NonRisposto
                    action.Esito = EsitoChiamata.NESSUNA_RISPOSTA
            End Select

            action.DettaglioEsito = c.EsitoChiamataEx
            action.Tag = c
            action.Ricevuta = False
            Select Case c.StatoChiamata
                Case StatoChiamataRegistrata.Sconosciuto, StatoChiamataRegistrata.InCorso, StatoChiamataRegistrata.Composizione
                    action.StatoConversazione = StatoConversazione.INATTESA
                Case Else
                    action.StatoConversazione = StatoConversazione.CONCLUSO
            End Select

            col.Add(action)
        End Sub

        Protected Overrides Sub FillSupportedTypes(items As CKeyCollection(Of String))
            items.Add("ChiamataRegistrata", "Chiamata Registrata")
        End Sub


    End Class


End Class